token-metadata/program/src/processor/burn/burn.rs

```diff
@@ -1,11 +1,195 @@
-use solana_program::{account_info::AccountInfo, entrypoint::ProgramResult, pubkey::Pubkey};
-
-use crate::{error::MetadataError, instruction::BurnArgs};
+use super::*;
+use crate::{
+    processor::burn::{fungible::burn_fungible, nonfungible_edition::burn_nonfungible_edition},
+    state::{AuthorityRequest, AuthorityType, TokenDelegateRole, TokenRecord, TokenState},
+    utils::{check_token_standard, thaw},
+};
 
 pub fn burn<'a>(
-    _program_id: &Pubkey,
-    _accounts: &'a [AccountInfo<'a>],
-    _args: BurnArgs,
+    program_id: &Pubkey,
+    accounts: &'a [AccountInfo<'a>],
+    args: BurnArgs,
 ) -> ProgramResult {
-    Err(MetadataError::FeatureNotSupported.into())
+    let context = Burn::to_context(accounts)?;
+
+    match args {
+        BurnArgs::V1 { .. } => burn_v1(program_id, context, args),
+    }
+}
+
+fn burn_v1(program_id: &Pubkey, ctx: Context<Burn>, args: BurnArgs) -> ProgramResult {
+    msg!("Burn V1");
+    let BurnArgs::V1 { amount } = args;
+
+    // Validate accounts
+
+    // Assert signer
+    assert_signer(ctx.accounts.authority_info)?;
+
+    // Assert program ownership.
+    assert_owned_by(ctx.accounts.metadata_info, program_id)?;
+    assert_owned_by(ctx.accounts.mint_info, &spl_token::ID)?;
+    assert_owned_by(ctx.accounts.token_info, &spl_token::ID)?;
+
+    if let Some(edition_info) = ctx.accounts.edition_info {
+        assert_owned_by(edition_info, program_id)?;
+    }
+
+    if let Some(parent_edition) = ctx.accounts.parent_edition_info {
+        assert_owned_by(parent_edition, program_id)?;
+    }
+    if let Some(parent_mint) = ctx.accounts.parent_mint_info {
+        assert_owned_by(parent_mint, &spl_token::ID)?;
+    }
+    if let Some(parent_token) = ctx.accounts.parent_token_info {
+        assert_owned_by(parent_token, &spl_token::ID)?;
+    }
+    if let Some(edition_marker) = ctx.accounts.edition_marker_info {
+        assert_owned_by(edition_marker, program_id)?;
+    }
```

Comment:
Agreed.

Reply:
Actually, adding token_record here makes sense, but since collection_metadata could be burned (and therefore owned by the system program) we want to wait to check ownership until after we check the burned case.